### PR TITLE
Simplex

### PR DESCRIPTION
**Pull request by brianpark0126**
Early push for prelim review. Still need to fix some comments, lint code, and test for small edge cases such as clipping inputs.

### CODE REVIEW

File: sotai/layers/lattice.py [hunk 1]
```diff
@@ -187,6 +186,109 @@ def _count_non_zeros(*iterables) -> int:
...
+        lattice_rank = len(self.lattice_sizes)
+        input_dim = len(inputs.shape)
+        all_size_2 = all(size == 2 for size in self.lattice_sizes)
+
+        # Strides are the index shift (with respect to flattened kernel data) of each
+        # dimension.
+        strides = torch.tensor(
+            np.cumprod([1] + self.lattice_sizes[::-1][:-1])[::-1].copy()
+        )
+        print(f"strides = {strides}")
```
Comment (willbakst) [inline sotai/layers/lattice.py:225]: nit: remove print statement

File: sotai/layers/lattice.py [hunk 2]
```diff
@@ -187,6 +186,109 @@ def _count_non_zeros(*iterables) -> int:
...
+        if all_size_2:
+            corner_offset_and_sorted_strides = torch.nn.functional.pad(
+                sorted_strides, [1, 0] + flat_no_padding
+            )
+        else:
+            corner_offset_and_sorted_strides = torch.cat(
+                [lower_corner_offset, sorted_strides], dim=-1
+            )
+        indices = torch.cumsum(corner_offset_and_sorted_strides, dim=-1)
+        print(f"indices: {indices}")
```
Comment (willbakst) [inline sotai/layers/lattice.py:272]: nit: remove print statement

File: sotai/layers/lattice.py [hunk 3]
```diff
@@ -187,6 +186,109 @@ def _count_non_zeros(*iterables) -> int:
...
+        else:
+            unit_offset = torch.tensor(
+                [[i] * (lattice_rank + 1) for i in range(self.units)]
+            )
+            flat_indices = indices * self.units + unit_offset
+            gathered_params = torch.index_select(
+                self.kernel.view(-1), 0, flat_indices.view(-1)
+            ).view(indices.shape)
+
+        # Dot product with interpolation weights. (todo: change)
```
Comment (willbakst) [inline sotai/layers/lattice.py:289]: what is the todo change here?

File: sotai/layers/lattice.py [hunk 4]
```diff
@@ -187,6 +186,109 @@ def _count_non_zeros(*iterables) -> int:
                 result += sum(1 for element in iterable if element != 0)
         return result
 
+    def _compute_simplex_interpolation(self, inputs: torch.Tensor) -> torch.Tensor:
+        """Evaluates a lattice using simplex interpolation.
+
+        Each unit hypercube of the lattice can be partitioned into d! disjoint simplices
```
Comment (willbakst) [inline sotai/layers/lattice.py:192]: this could use some freshening up / grammar fixes

File: sotai/layers/lattice.py [hunk 5]
```diff
@@ -187,6 +186,109 @@ def _count_non_zeros(*iterables) -> int:
...
+
+        if self.clip_inputs:
+            inputs = self._clip_onto_lattice_range(inputs)
+
+        lattice_rank = len(self.lattice_sizes)
+        input_dim = len(inputs.shape)
+        all_size_2 = all(size == 2 for size in self.lattice_sizes)
+
+        # Strides are the index shift (with respect to flattened kernel data) of each
+        # dimension.
```
Comment (willbakst) [inline sotai/layers/lattice.py:221]: maybe flesh this out further to describe in detail what the index math you're doing is actually doing with an example potentially

File: sotai/layers/lattice.py [hunk 6]
```diff
@@ -187,6 +186,119 @@ def _count_non_zeros(*iterables) -> int:
...
+                [lower_corner_offset, sorted_strides], dim=-1
+            )
+        indices = torch.cumsum(corner_offset_and_sorted_strides, dim=-1)
+
+        # Get kernel data from corresponding simplex vertices.
+        if self.units == 1:
+            gathered_params = torch.index_select(
+                self.kernel.view(-1), 0, indices.view(-1)
+            ).view(indices.shape)
+            print(f"gathered_params: {gathered_params}")
```
Comment (willbakst) [inline sotai/layers/lattice.py:288]: nit: remove print statement here

**Review by willbakst — APPROVED** [review]
